File: app/src/main/java/hr/caellian/notestream/gui/fragments/FragmentPlayableTile.java
package hr.caellian.notestream.gui.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import hr.caellian.notestream.NoteStream;
import hr.caellian.notestream.R;
import hr.caellian.notestream.data.playable.PlayableRemote;
import hr.caellian.notestream.gui.ActivityPlayer;
import hr.caellian.notestream.data.PlayerService;
import hr.caellian.notestream.data.Playlist;
import hr.caellian.notestream.data.playable.Playable;
import hr.caellian.notestream.gui.PlayablePopupMenu;
import hr.caellian.notestream.lib.Constants;

/**
 * Created by caellyan on 21/06/17.
 */

public class FragmentPlayableTile extends FragmentPlayableMediator {

    public static FragmentPlayableTile newInstance(Playable playable, Playlist playlist) {
        FragmentPlayableTile fragment = new FragmentPlayableTile();

        Bundle args = new Bundle();
        args.putString(Constants.EXTRA_PLAYLIST, playlist.getID());
        args.putString(Constants.EXTRA_PLAYABLE, playable.getID());
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.playable_tile, container, false);
        defaultTextColor = ((TextView) view.findViewById(R.id.labelSongTitle)).getCurrentTextColor();

        Playable playable = getPlayable();

        if (playable instanceof PlayableRemote && !((PlayableRemote) playable).isAvailable()) {
            NoteStream.registerAvailabilityListener(this);
            view.setVisibility(View.GONE);
        } else {
            updateView();
        }

        return view;
    }

    protected void updateView() {
        final Playlist playlist = getPlaylist();
        final Playable playable = getPlayable();

        if (playable == null) {
            return;
        }

        ((ImageView) view.findViewById(R.id.tileBG)).setImageBitmap(playable.getMetadata().getCover());
        ((TextView) view.findViewById(R.id.labelSongTitle)).setText(playable.getMetadata().getTitle());
        ((TextView) view.findViewById(R.id.labelSongAuthor)).setText(playable.getMetadata().getAuthor());

        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (playable instanceof PlayableRemote && !((PlayableRemote) playable).isAvailable()) {
                    return;
                }

                PlayerService.PlayerServiceBinder psb = NoteStream.getInstance().getPlayerServiceBinder();
                if (psb.getCurrentPlayable() != playable) {
                    psb.playAt(playlist, playable);
                } else {
                    psb.setPlaylist(playlist);
                }

                Intent intent = new Intent(view.getContext(), ActivityPlayer.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
                intent.putExtra(Constants.EXTRA_PLAYLIST, playlist.getID());
                intent.putExtra(Constants.EXTRA_PLAYABLE, playable.getID());
                startActivity(intent);
            }
        });

        view.setOnLongClickListener(new View.OnLongClickListener() {
            PlayablePopupMenu playablePopupMenu = new PlayablePopupMenu(view.getContext(), view, playable);

            @Override
            public boolean onLongClick(View v) {
                playablePopupMenu.show();
                return true;
            }
        });

        NoteStream.registerControlListener(this);
        NoteStream.registerLibraryListener(this);
    }
}

File: app/src/main/java/hr/caellian/notestream/gui/fragments/FragmentPlayableTile.kt
/*
 * Copyright (C) 2018 Tin Svagelj
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
 * persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
 * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package hr.caellian.notestream.gui.fragments

import android.content.Intent
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import hr.caellian.notestream.NoteStream
import hr.caellian.notestream.NoteStream.PlayerServiceListener
import hr.caellian.notestream.R
import hr.caellian.notestream.data.PlayerService
import hr.caellian.notestream.data.playable.Playable
import hr.caellian.notestream.data.playable.PlayableRemote
import hr.caellian.notestream.data.playlist.Playlist
import hr.caellian.notestream.gui.ActivityPlayer
import hr.caellian.notestream.gui.PlayablePopupMenu
import hr.caellian.notestream.lib.Constants

class FragmentPlayableTile : FragmentPlayableMediator(LinearLayout.VERTICAL) {
    private var defaultTextColor: Int = 0

    override fun inflateContentView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val view = inflater.inflate(R.layout.fragment_tile, container, false)
        defaultTextColor = view.findViewById<TextView>(R.id.labelTileTitle)?.currentTextColor ?: 0

        return view
    }

    override fun updateView(rootView: View) {
        val playlist = argumentPlaylist ?: return
        val playable = argumentPlayable ?: return

        rootView.findViewById<ImageView>(R.id.imageTileBackground)?.setImageBitmap(playable.info.cover)
        rootView.findViewById<TextView>(R.id.labelTileTitle)?.text = playable.title
        rootView.findViewById<TextView>(R.id.labelSongAuthor)?.text = playable.author

        rootView.setOnClickListener(View.OnClickListener {
            if (playable is PlayableRemote && !playable.available) {
                return@OnClickListener
            }

            NoteStream.registerPlayerServiceListener(object : PlayerServiceListener() {
                override fun onPlayerServiceConnected(psb: PlayerService.PlayerServiceBinder) {
                    if (psb.currentPlayable != playable) {
                        psb.playAt(playlist, playable)
                    } else {
                        psb.setPlaylist(playlist)
                    }
                }
            })

            val intent = Intent(rootView.context, ActivityPlayer::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT)
            intent.putExtra(Constants.EXTRA_PLAYLIST, playlist.id)
            intent.putExtra(Constants.EXTRA_PLAYABLE, playable.id)
            startActivity(intent)
        })

        rootView.setOnLongClickListener {
            val playablePopupMenu = PlayablePopupMenu(it.context, it, playable)
            playablePopupMenu.show()
            true
        }

        NoteStream.registerControlListener(this)
        NoteStream.registerLibraryListener(this)
    }

    override fun onPlayableChanged(current: Playable?) {
        if (current == argumentPlayable) {
            view?.findViewById<TextView>(R.id.labelTileTitle)?.setTextColor(ContextCompat.getColor(view!!.context, R.color.colorAccent))
        } else {
            view?.findViewById<TextView>(R.id.labelTileTitle)?.setTextColor(defaultTextColor)
        }
    }

    companion object {
        fun create(playlist: Playlist, playable: Playable): FragmentPlayableTile {
            val fragment = FragmentPlayableTile()

            val args = Bundle()
            args.putString(Constants.EXTRA_PLAYLIST, playlist.id)
            args.putString(Constants.EXTRA_PLAYABLE, playable.id)
            fragment.arguments = args
            return fragment
        }
    }
}
